libraries/standard/mqtt/src/iot_mqtt_serialize.c

```diff
@@ -465,6 +486,141 @@ static uint8_t * _encodeString( uint8_t * pDestination,
+            UINT8_SET_BIT( connectFlags, MQTT_CONNECT_FLAG_USERNAME );
+        #endif
+    }
+    else
+    {
+        /* Set the flags for username and password if provided. */
+        if( pConnectInfo->pUserName != NULL )
+        {
+            UINT8_SET_BIT( connectFlags, MQTT_CONNECT_FLAG_USERNAME );
+        }
+
+        if( pConnectInfo->pPassword != NULL )
+        {
+            UINT8_SET_BIT( connectFlags, MQTT_CONNECT_FLAG_PASSWORD );
+        }
+    }
+
+    /* Set will flag if an LWT is provided. */
+    if( pConnectInfo->pWillInfo != NULL )
+    {
+        UINT8_SET_BIT( connectFlags, MQTT_CONNECT_FLAG_WILL );
+
+        /* Flags only need to be changed for will QoS 1 and 2. */
+        switch( pConnectInfo->pWillInfo->qos )
+        {
+            case IOT_MQTT_QOS_1:
+                UINT8_SET_BIT( connectFlags, MQTT_CONNECT_FLAG_WILL_QOS1 );
+                break;
+
+            case IOT_MQTT_QOS_2:
+                UINT8_SET_BIT( connectFlags, MQTT_CONNECT_FLAG_WILL_QOS2 );
+                break;
+
+            default:
+                break;
+        }
+
+        if( pConnectInfo->pWillInfo->retain == true )
+        {
+            UINT8_SET_BIT( connectFlags, MQTT_CONNECT_FLAG_WILL_RETAIN );
+        }
+    }
+    return connectFlags;
+}
+
+/*-----------------------------------------------------------*/
+
+static uint8_t * _encodeUserName( uint8_t * pBuffer,
+                                  const IotMqttConnectInfo_t * pConnectInfo )
+{
+    bool encodedUserName = false;
+
+    /* If metrics are enabled, write the metrics username into the CONNECT packet.
+     * Otherwise, write the username and password only when not connecting to the
+     * AWS IoT MQTT server. */
+    if( pConnectInfo->awsIotMqttMode == true )
+    {
+        #if AWS_IOT_MQTT_ENABLE_METRICS == 1
+            IotLogInfo( "Anonymous metrics (SDK language, SDK version) will be provided to AWS IoT. "
+                        "Recompile with AWS_IOT_MQTT_ENABLE_METRICS set to 0 to disable." );
+
+            /* Determine if the Connect packet should use a combination of the username
+             * for authentication plus the SDK version string. */
+            if( pConnectInfo->pUserName != NULL )
+            {
+                /* Only include metrics if it will fit within the encoding
+                 * standard. */
+                if( ( pConnectInfo->userNameLength + AWS_IOT_METRICS_USERNAME_LENGTH ) <= UINT16_MAX )
+                {
+                    /* Write the high byte of the combined length. */
+                    *( pBuffer++ ) = UINT16_HIGH_BYTE( ( pConnectInfo->userNameLength +
+                                                         AWS_IOT_METRICS_USERNAME_LENGTH ) );
+
+                    /* Write the low byte of the combined length. */
+                    *( pBuffer++ ) = UINT16_LOW_BYTE( ( pConnectInfo->userNameLength +
+                                                        AWS_IOT_METRICS_USERNAME_LENGTH ) );
```

Comment:
Originally this was:
```
*pBuffer = ...
*( pBuffer + sizeof( uint8_t ) ) = ...
pBuffer += sizeof( uint16_t );
```

The `sizeof`s can be changed to `1` and `2`, but the complexity of this function would still be 8. Changing those three lines to two `*(pBuffer++)`'s reduced the complexity to 7, probably since there are less lines of code.

Reply:
Actually this looks rather confusing since UINT16_* are macros already.  Maybe change it back.

Reply:
I'm not sure I understand. The only changes were:
`*pBuffer` -> `*( pBuffer++ )`
`*( pBuffer + sizeof( uint8_t ) )` -> `*( pBuffer++ )`
`pBuffer += sizeof( uint16_t );` -> deleted.

The UINT16_* were there already. Not making this change leaves the complexity at 8.